### PR TITLE
[Tune] Add repr for ResultGrid class

### LINKED ISSUE #31716: [Tune] Improve `ResultGrid` repr to make its accessible attributes more obvious
### Description

One possibility:

```python
>>> result_grid: ResultGrid = tuner.fit()
>>> result_grid
ResultGrid([
    Result(metrics=..., log_dir=..., checkpoint=...),
    Result(metrics=..., log_dir=..., checkpoint=...),
])
```

### Use case

I want to work with a `ResultGrid` interactively in a Jupyter notebook. The repr would ideally help me understand that this is an iterable set of Result objects that I can work with individually.

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?
Add __repr__() for ResultGrid class and prettify __repr__() of Result class.

The example representation:

```python
>>> result_grid: ResultGrid = tuner.fit()
>>> result_grid
[
  Result(
    error='RayTaskError(RuntimeError)',
    metrics={'trial_id': '820f5_00000'},
    log_dir=PosixPath('/private/var/folders/0w/7cw9tmnx6nsfxtknpt99s7gc0000gn/T/pytest-of-yunxuanx/pytest-18/test_result_grid_repr0/ray_results/test_GridResults_repr/f_820f5_00000_0_x=1_2023-01-26_17-18-30'),
    checkpoint=None
  ),
  Result(
    metrics={'loss': 1, '_metadata': _CheckpointMetadata(checkpoint_type=<class 'ray.air.checkpoint.Checkpoint'>, checkpoint_state={}), 'should_checkpoint': True, 'done': True, 'trial_id': '820f5_00001', 'experiment_tag': '1_x=2'},
    log_dir=PosixPath('/private/var/folders/0w/7cw9tmnx6nsfxtknpt99s7gc0000gn/T/pytest-of-yunxuanx/pytest-18/test_result_grid_repr0/ray_results/test_GridResults_repr/f_820f5_00001_1_x=2_2023-01-26_17-18-31'),
    checkpoint=Checkpoint(local_path=/private/var/folders/0w/7cw9tmnx6nsfxtknpt99s7gc0000gn/T/pytest-of-yunxuanx/pytest-18/test_result_grid_repr0/ray_results/test_GridResults_repr/f_820f5_00001_1_x=2_2023-01-26_17-18-31/checkpoint_000000)
  ),
]
```

## Related issue number
Closes https://github.com/ray-project/ray/issues/31716

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/test_result_grid.py [hunk 1]
```diff
@@ -224,6 +224,26 @@ def f(config):
     assert not any(key in representation for key in AUTO_RESULT_KEYS)
 
 
+def test_result_grid_repr(ray_start_2_cpus):
+    def f(config):
+        with tune.checkpoint_dir(step=0) as checkpoint_dir:
+            path = os.path.join(checkpoint_dir, "checkpoint")
+            with open(path, "w") as f:
+                f.write(json.dumps({"step": 0}))
+
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:233]: We can save a checkpiont with `session.report`. This `tune.checkpoint_dir` was the old API.
```suggestion
```

File: python/ray/tune/tests/test_result_grid.py [hunk 2]
```diff
@@ -224,6 +224,26 @@ def f(config):
...
+        from ray.air import session
+
+        session.report({"loss": 1})
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:236]: Also make sure the `Checkpoint` gets imported with `from ray.air import Checkpoint`

```suggestion
        session.report({"loss": 1}, checkpoint=Checkpoint.from_dict({"step": 0}))
```

File: python/ray/tune/tests/test_result_grid.py [hunk 3]
```diff
@@ -224,6 +224,26 @@ def f(config):
...
+
+    tuner = tune.Tuner(f, param_space={"x": tune.grid_search([1, 2])})
+    result_grid = tuner.fit()
+
+    from ray.tune.result import AUTO_RESULT_KEYS
+
+    representation = result_grid.__repr__()
+    assert not any(key in representation for key in AUTO_RESULT_KEYS)
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:244]: Can we check that 2 result reprs get shown in the result grid repr? Checking that all trials get printed?

File: python/ray/tune/tests/test_result_grid.py [hunk 4]
```diff
@@ -224,6 +224,26 @@ def f(config):
     assert not any(key in representation for key in AUTO_RESULT_KEYS)
 
 
+def test_result_grid_repr(ray_start_2_cpus):
+    def f(config):
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:228]: We can actually mock an error for one of the trials, then can test that it shows up in the repr later:

```python
if config["x"] == 1:
    raise RuntimeError
```

**Review by justinvyu — CHANGES_REQUESTED** [review]
Looks good! Just a few minor change suggestions.

File: python/ray/tune/tests/test_result_grid.py (hunk 4 above)
Comment (woshiyyya) [reply python/ray/tune/tests/test_result_grid.py:228]: I see. So the error thrown inside the trainable will be captured and stored in the Result object right?

File: python/ray/tune/tests/test_result_grid.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/tune/tests/test_result_grid.py:236]: Yes, it has been imported at the beginning.

File: python/ray/tune/tests/test_result_grid.py (hunk 4 above)
Comment (justinvyu) [reply python/ray/tune/tests/test_result_grid.py:228]: Yep!

**Review by justinvyu — APPROVED** [review]
Looks good!

File: python/ray/tune/tests/test_result_grid.py [hunk 5]
```diff
@@ -224,6 +224,39 @@ def f(config):
...
+        f,
+        run_config=air.RunConfig(
+            name="test_result_grid_repr",
+            local_dir=str(tmpdir / "test_result_grid_repr_results"),
+        ),
+        param_space={"x": tune.grid_search([1, 2])},
+    )
+    result_grid = tuner.fit()
+
+    representation = result_grid.__repr__()
```
Comment (krfricke) [inline python/ray/tune/tests/test_result_grid.py:247]: Can we instead construct the result grid manually and assert the layout?

File: python/ray/air/result.py [hunk 6]
```diff
@@ -57,10 +57,16 @@ def __repr__(self):
...
 
         if self.metrics:
             shown_attributes["metrics"] = {
                 k: v for k, v in self.metrics.items() if k not in AUTO_RESULT_KEYS
             }
+
         kws = [f"{key}={value!r}" for key, value in shown_attributes.items()]
-        return "{}({})".format(type(self).__name__, ", ".join(kws))
+        kws_repr = "  " + ",\n  ".join(kws)
+        return "{}(\n{}\n)".format(type(self).__name__, kws_repr)
```
Comment (krfricke) [inline python/ray/air/result.py:72]: It's very hard to imagine how this looks like, can we add a comment what we expect here?

File: python/ray/tune/result_grid.py [hunk 7]
```diff
@@ -240,3 +240,10 @@ def _trial_to_result(self, trial: Trial) -> Result:
             best_checkpoints=best_checkpoints,
         )
         return result
+
+    def __repr__(self) -> str:
+        all_results_repr = ""
+        for result in self:
+            result_repr = "  " + result.__repr__().replace("\n", "\n  ")
+            all_results_repr += f"{result_repr},\n"
```
Comment (krfricke) [inline python/ray/tune/result_grid.py:248]: 1. Can we use a list comprehension here?
2. The replace looks a bit fishy. Can we maybe instead use a private method in `Result` for both `Result.__repr__` and `ResultGrid.__repr__`?

File: python/ray/tune/result_grid.py [hunk 8]
```diff
@@ -240,3 +240,10 @@ def _trial_to_result(self, trial: Trial) -> Result:
...
+        return f"[\n{all_results_repr}]"
```
Comment (krfricke) [inline python/ray/tune/result_grid.py:249]: Should we still have `ResultGrid<[ ... ]>` in here?

**Review by krfricke — CHANGES_REQUESTED** [review]
I have a few minor questions for the repr implementation.

For testing, this is currently and end to end test, but I don't think this is needed. We can just create Result objects manually and a ResultGrid containing these elements. This will lead to a test execution time of a few milliseconds compared to the few seconds it takes now to run the tune experiment.

Generally we should aim to use minimalistic tests if possible - we don't get any benefit from a full end to end test here, so we should just stick with a simple unit test.

File: python/ray/tune/tests/test_result_grid.py (hunk 5 above)
Comment (woshiyyya) [reply python/ray/tune/tests/test_result_grid.py:247]: I agree. Manually creating a result grid will be way faster than running a full tuning experiment.

File: python/ray/tune/tests/test_result_grid.py [hunk 9]
```diff
@@ -224,6 +224,62 @@ def f(config):
     assert not any(key in representation for key in AUTO_RESULT_KEYS)
 
 
+def test_result_grid_repr(ray_start_2_cpus, tmpdir):
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:227]: No longer need to do ray init if we create the result objects directly.
```suggestion
def test_result_grid_repr(tmpdir):
```

File: python/ray/tune/tests/test_result_grid.py [hunk 10]
```diff
@@ -224,6 +224,62 @@ def f(config):
...
+    from ray.air.result import Result
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:228]: Can move this to the top imports.

File: python/ray/tune/tests/test_result_grid.py [hunk 11]
```diff
@@ -224,6 +224,62 @@ def f(config):
...
+
+    results = list()
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:230]: Nit: We can just construct the full list directly, like:
```python
result_grid._results = [
    Result(...),
    Result(...),
]
```

File: python/ray/tune/result_grid.py [hunk 12]
```diff
@@ -69,6 +69,13 @@ def __init__(
         experiment_analysis: ExperimentAnalysis,
     ):
         self._experiment_analysis = experiment_analysis
+        self._results = list()
+
+        if experiment_analysis:
```
Comment (justinvyu) [inline python/ray/tune/result_grid.py:74]: `experiment_analysis` is a non-optional parameter, so we can just always do the thing in the if statement.

File: python/ray/tune/result_grid.py [hunk 13]
```diff
@@ -69,6 +69,13 @@ def __init__(
...
+            self._results = [
+                self._trial_to_result(trail)
+                for trail in self._experiment_analysis.trials
```
Comment (justinvyu) [inline python/ray/tune/result_grid.py:77]:
```suggestion
                self._trial_to_result(trial)
                for trial in self._experiment_analysis.trials
```

File: python/ray/air/result.py [hunk 14]
```diff
@@ -53,14 +53,29 @@ def config(self) -> Optional[Dict[str, Any]]:
             return None
         return self.metrics.get("config", None)
 
-    def __repr__(self):
+    def _repr_with_indent(self, offset: int = 0) -> str:
```
Comment (justinvyu) [inline python/ray/air/result.py:56]: Nit: 
```suggestion
    def _repr(self, indent: int = 0) -> str:
```

File: python/ray/tune/result_grid.py [hunk 15]
```diff
@@ -240,3 +245,8 @@ def _trial_to_result(self, trial: Trial) -> Result:
             best_checkpoints=best_checkpoints,
         )
         return result
+
+    def __repr__(self) -> str:
+        all_results_repr = [result._repr_with_indent(2) for result in self]
```
Comment (justinvyu) [inline python/ray/tune/result_grid.py:250]:
```suggestion
        all_results_repr = [result._repr_with_indent(offset=2) for result in self]
```

File: python/ray/tune/tests/test_result_grid.py [hunk 16]
```diff
@@ -224,6 +224,62 @@ def f(config):
...
+    results.append(
+        Result(
+            metrics={"loss": 1.0},
+            checkpoint=Checkpoint(data_dict={"weight": 1.0}),
+            log_dir=Path("./log_1"),
+            error=None,
+            metrics_dataframe=None,
+            best_checkpoints=None,
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:238]:
```suggestion
```

File: python/ray/tune/tests/test_result_grid.py [hunk 17]
```diff
@@ -224,6 +224,62 @@ def f(config):
...
+        )
+    )
+    results.append(
+        Result(
+            metrics={"loss": 2.0},
+            checkpoint=Checkpoint(data_dict={"weight": 2.0}),
+            log_dir=Path("./log_2"),
+            error=RuntimeError(),
+            metrics_dataframe=None,
+            best_checkpoints=None,
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:248]:
```suggestion
```

File: python/ray/tune/tests/test_result_grid.py [hunk 18]
```diff
@@ -224,6 +224,62 @@ def f(config):
...
+    result_grid._results = results
+
+    representation = result_grid.__repr__()
+
+    from ray.tune.result import AUTO_RESULT_KEYS
+
+    assert len(result_grid) == 2
+    assert not any(key in representation for key in AUTO_RESULT_KEYS)
+
+    gold_representation = """ResultGrid<[
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:262]:
```suggestion
    expected_repr = """ResultGrid<[
```

File: python/ray/tune/tests/test_result_grid.py [hunk 19]
```diff
@@ -224,6 +224,62 @@ def f(config):
...
+    log_dir=PosixPath('log_2'),
+    checkpoint=Checkpoint(data_dict={'weight': 2.0})
+  )
+]>"""
+
+    assert representation == gold_representation
+    assert representation.count("metrics=") == 2
+    assert representation.count("log_dir=") == 2
+    assert representation.count("checkpoint=") == 2
+    assert representation.count("error=") == 1 and "RuntimeError" in representation
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:280]: Nit: Maybe we don't need these anymore.

File: python/ray/tune/tests/test_result_grid.py [hunk 20]
```diff
@@ -224,6 +224,62 @@ def f(config):
...
+            metrics={"loss": 2.0},
+            checkpoint=Checkpoint(data_dict={"weight": 2.0}),
+            log_dir=Path("./log_2"),
+            error=RuntimeError(),
+            metrics_dataframe=None,
+            best_checkpoints=None,
+        )
+    )
+
+    result_grid = ResultGrid(experiment_analysis=None)
```
Comment (justinvyu) [inline python/ray/tune/tests/test_result_grid.py:252]: Ah, I see why we had the check if `experiment_analysis` is `None` now. We can just create a mock analysis object that contains an empty list of trials. This way, we don't need to change the experiment analysis to be an optional parameter.

```python
class MockExperimentAnalysis:
    trials = []
```

**Review by justinvyu — CHANGES_REQUESTED** [review]
I have a few more small suggestions, but should be good after that! The new `_results` is definitely better, and +1 to creating the result objects directly in the test without running an actual experiment.

File: python/ray/tune/tests/test_result_grid.py (hunk 17 above)
Comment (justinvyu) [reply python/ray/tune/tests/test_result_grid.py:249]: Actually, these are required arguments. Closing.

**Review by justinvyu — APPROVED** [review]
LGTM!

**Review by krfricke — APPROVED** [review]
Nice! There's one failing test please ping me once resolved and CI is passing

**Pull request merged**